### PR TITLE
[LLD][COFF] Introduce hybrid symbol table for EC input files on ARM64X

### PR DESCRIPTION
**Pull request by cjacek**
On hybrid ARM64X targets, ARM64 and ARM64EC input files operate in separate namespaces and cannot reference each other. This change introduces separate `SymbolTable` instances and associates each `InputFile` with the appropriate table to reflect this behavior.

### CODE REVIEW

**Comment by cjacek** [comment]
This is the beginning of a separate namespace implementation required for hybrid ARM64X image support. This change introduces just enough functionality to handle both native and EC load configs passed to the linker. This is how it should work, but there are still many missing pieces. In particular, CHPE metadata should only be present in the EC version and copied over to the native load config. I’ve created an [arm64x-loadcfg](https://github.com/cjacek/llvm-project/commits/arm64x-loadcfg) branch with additional ARM64X load config fixes to provide a clearer context. For now, I build the native load config from `loadconfig-arm64ec.s` in tests, which also contains CHPE metadata. To keep it working, I continue to use `ctx.config.machine` for the condition defining `__arm64x_*` and `__hybrid_*` symbols. Ultimately, this should use `symtab.isEC()`, like it’s done in the arm64x-loadcfg branch, but that requires additional changes. I kept this initial version simple to facilitate incremental development.

A lot of existing code assumes a single symbol table. Most of the remaining ARM64X changes will involve:
- Modifying the code to use `symtab` instead of `ctx` in cases that need to operate on a specific symbol table (similar to how it’s done for `InputFile` here).
- Modifying the code to use `ctx.forEachSymtab` in cases where the same operation needs to be performed on all symbol tables.
- Using `ctx.symtabEC` for code specific to the EC symbol table (many of these changes are part of the arm64x-loadcfg branch).

I will create separated PRs for changes preparing for the final commit.

File: lld/COFF/Chunks.cpp [hunk 1]
```diff
@@ -129,7 +129,7 @@ void SectionChunk::applyRelX64(uint8_t *off, uint16_t type, OutputSection *os,
   case IMAGE_REL_AMD64_REL32_4:  add32(off, s - p - 8); break;
   case IMAGE_REL_AMD64_REL32_5:  add32(off, s - p - 9); break;
   case IMAGE_REL_AMD64_SECTION:
-    applySecIdx(off, os, file->ctx.outputSections.size());
+    applySecIdx(off, os, file->symtab.ctx.outputSections.size());
```
Comment (MaskRay) [inline lld/COFF/Chunks.cpp:132]: drive-by: why `file->symtab.ctx` instead of `file->ctx`?

Comment (cjacek) [reply lld/COFF/Chunks.cpp:132]: This change stems from addressing the opposite scenario: determining the correct symtab from the input file. With hybrid images and two namespaces, `file->ctx.symtab` isn’t always accurate. EC files should use the hybrid symbol table instead, which would look something like `file->ctx.getSymtab(file->getMachineType())` in this PR.

My initial draft followed this approach (encapsulated in an `InputFile` helper) and updated all relevant code to utilize it. However, this approach felt suboptimal - it distributed complexity across the codebase and increased the risk of mistakes or omissions. Instead, storing `symtab` directly in `InputFile` and determining the correct symbol table during object construction made it easier to access and encapsulated the complexity.

I considered storing both `COFFLinkerContext` and `SymbolTable` references in the `InputFile` class. This would allow both `file->ctx` and `file->symtab` to work. However, since the context is easily accessible from the symbol table, this approach felt redundant. To save memory, I opted to store only `symtab`, replacing `file->ctx` with `file->symtab.ctx` as seen in #119296.

The required changes to use `file->symtab.ctx` seemed manageable to me. Additionally, with further refactoring, the number of such instances will likely decrease. For instance, commit 48b055dbd4961f4c in #119298 removes many such cases. Another example would be updating `ImportThunkChunk` to store `symtab` instead of `ctx`, further reducing these instances.

**Comment by cjacek** [comment]
NFC refactoring is done is separate PRs, this depends on #119298.

File: lld/COFF/Driver.cpp [hunk 2]
```diff
@@ -2519,54 +2529,56 @@ void LinkerDriver::linkerMain(ArrayRef<const char *> argsArr) {
...
+    symtab.addAbsolute(mangle("__guard_iat_table"), 0);
+    symtab.addAbsolute(mangle("__guard_longjmp_count"), 0);
+    symtab.addAbsolute(mangle("__guard_longjmp_table"), 0);
+    // Needed for MSVC 2017 15.5 CRT.
+    symtab.addAbsolute(mangle("__enclave_config"), 0);
+    // Needed for MSVC 2019 16.8 CRT.
+    symtab.addAbsolute(mangle("__guard_eh_cont_count"), 0);
+    symtab.addAbsolute(mangle("__guard_eh_cont_table"), 0);
+
+    if (isArm64EC(ctx.config.machine)) {
```
Comment (mstorsjo) [inline lld/COFF/Driver.cpp:2552]: I guess this bit, about what symbols we set in the load config, is going to change further in some subsequent change?

File: lld/test/COFF/arm64x-undef.s [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+// RUN: llvm-mc -filetype=obj -triple=arm64ec-windows sym.s -o sym-arm64ec.obj
+// RUN: llvm-mc -filetype=obj -triple=x86_64-windows sym.s -o sym-x86_64.obj
+// RUN: llvm-mc -filetype=obj -triple=aarch64-windows symref.s -o symref-aarch64.obj
+
+// Check that native object files can't reference EC symbols.
+
+// RUN: not lld-link -machine:arm64x -dll -noentry -out:out.dll symref-aarch64.obj sym-arm64ec.obj \
+// RUN:              2>&1 | FileCheck --check-prefix=UNDEF %s
+// UNDEF:      lld-link: error: undefined symbol: sym
```
Comment (mstorsjo) [inline lld/test/COFF/arm64x-undef.s:12]: I think this test would be clearer if we'd also have a positive test, to see that symbols within each namespace can get resolved correctly. I guess that's kinda implicitly tested in all the other existing tests, but it would be nice for clarity here as well.

**Review by mstorsjo — COMMENTED** [review]
Overall this looks very straightforward, almost suspiciously straightforward :-)

Can you verbally explain what the state is after this commit? We essentially do symbol resolution separately within the both namespaces - and when writing the final image, we just serialize all included Chunks into the output image?

What's the reason why we now link two separate copies of the load config in the testcases - what difference does that make? (If we wouldn't be linking two load configs but run the old testcases as they were, what would the results be?)

**Comment by cjacek** [comment]
Thanks for reviews!

> Can you verbally explain what the state is after this commit? We essentially do symbol resolution separately within the both namespaces - and when writing the final image, we just serialize all included Chunks into the output image?

Yes, with this commit, we now have two distinct namespaces, each with independent name resolution. The writer still operates on file and/or chunk collections globally rather than being tied to a specific namespace. Its existing functionality remains unchanged, as it already handles different chunk machine types, including sorting, splitting, and aligning code chunks. Those parts are largely complete.

Most of the writer and driver code still uses `ctx.symtab` and ignores the hybrid symbol table. While this is correct in some cases (e.g., the PE header writer only needs to operate on the native namespace), other parts will require adjustments as outlined in the first comment.

> What's the reason why we now link two separate copies of the load config in the testcases - what difference does that make? (If we wouldn't be linking two load configs but run the old testcases as they were, what would the results be?)

PE header writer uses the native namespace to look up the load config referenced by the PE data directory. Without adding the second load config, executables created by the tests would lack a load config entirely. The hybrid load config is only accessible after applying ARM64X relocations, but those relocations are exposed to the loader via the native load config itself. Without a native load config, they wouldn't be visible (load config data directory ARM64X relocations aren't part of this patch; see [this commit](https://github.com/cjacek/llvm-project/commit/e7f8cacdcfbcced24f1985ce132aadd2cbe9cf2f) for how I plan to implement that part).

For this commit, replacing the EC load config with the native one might work but wouldn't represent real progress. Ultimately, we should pass both load configs. The current hack, which allowed for splitting commits, involves providing CHPE metadata in the native load config. This metadata should instead be present only in the EC version, which we then copy over by the linker to the native version. See [this commit ](https://github.com/cjacek/llvm-project/commit/990b779ec6897fdd51dd5b1a13df085b390ccf2d) for an implementation that aligns tests with the intended use of load configs on ARM64X.

File: lld/COFF/Driver.cpp (hunk 2 above)
Comment (cjacek) [reply lld/COFF/Driver.cpp:2552]: Yes, this should ultimately use `symtab.isEC()` instead. Changing it now would break CHPE data in the native load config. CHPE metadata should only exist in the EC namespace, but maintaining compatibility with existing tests requires linker support to copy it from the EC load config to the native one, as well as updates to the writer to correctly set those symbols in the appropriate namespace. This is implemented in [this commit](https://github.com/cjacek/llvm-project/commit/990b779ec6897fdd51dd5b1a13df085b390ccf2d).

File: lld/test/COFF/arm64x-undef.s (hunk 3 above)
Comment (cjacek) [reply lld/test/COFF/arm64x-undef.s:12]: Good idea, I'll add such a test.

Comment (cjacek) [reply lld/test/COFF/arm64x-undef.s]: Done. I also added a test reading both ECSYMBOLS and native symbol indexes from an archive.

**Review by mstorsjo — APPROVED** [review]
LGTM, thanks!

**Comment by cjacek** [comment]
I reverted the commit due to sanitizer test failures.
